include/tvm/runtime/crt/rpc_common/framing.h

```diff
@@ -0,0 +1,267 @@
+ public:
+  explicit Unframer(WriteStream* stream)
+      : stream_{stream},
+        state_{State::kFindPacketStart},
+        saw_escape_start_{false},
+        num_buffer_bytes_valid_{0} {}
+
+  /*!
+   * \brief Push data into unframer and try to decode one packet.
+   *
+   * This function will return when exactly one packet has been decoded. It may not consume all of
+   * `data` in this case, and valid bytes may remain at the end of data.
+   *
+   * \param data The new data to unframe and send downstream.
+   * \param data_size_bytes The number of valid bytes in data.
+   * \param bytes_consumed Pointer written with the number of bytes consumed from data.
+   * \return
+   *     - kTvmErrorNoError when successful -- continue writing data.
+   *     - kTvmErrorFramingInvalidState when the Unframer was in or enters an invalid state
+   *       (probably indicates memory corruption).
+   *     - kTvmErrorFramingShortPacket when a new packet started before the current one ended.
+   *     - kTvmErrorFramingInvalidEscape when an invalid escape sequence was seen
+   */
+  tvm_crt_error_t Write(const uint8_t* data, size_t data_size_bytes, size_t* bytes_consumed);
+
+  /*! \brief Reset unframer to initial state. */
+  void Reset();
+
+  /*! \brief Return an underestimate of the number of bytes needed from the wire. */
+  size_t BytesNeeded();
+
+ private:
+  tvm_crt_error_t FindPacketStart();
+  tvm_crt_error_t FindPacketLength();
+  tvm_crt_error_t FindPacketCrc();
+  tvm_crt_error_t FindCrcEnd();
+
+  inline bool is_buffer_full(size_t buffer_full_bytes) {
+    return num_buffer_bytes_valid_ >= buffer_full_bytes;
+  }
+
+  /*! \brief Consume input into buffer_ until buffer_ has buffer_full_bytes. */
+  tvm_crt_error_t AddToBuffer(size_t buffer_full_bytes, bool update_crc);
+
+  void ClearBuffer();
+
+  /*! \brief Unescape and consume input bytes, storing into buffer.
+   *
+   * \param buffer A buffer to fill with consumed, unescaped bytes.
+   * \param buffer_size_bytes Size of buffer, in bytes.
+   * \param bytes_filled A pointer to an accumulator to which is added the number of bytes written
+   *      to `buffer`.
+   * \param update_crc true when the CRC should be updated with the escaped bytes.
+   * \return
+   *     - kTvmErrorNoError if successful
+   *     - kTvmErrorFramingShortPacket if a start-of-packet escape code was encountered. If so,
+   *       *bytes_filled indicates the number of bytes before the Escape::kEscapeStart byte.
+   *     - kTvmErrorFramingInvalidEscape if an invalid escape sequence was seen.
+   *     - kTvmErrorWriteStreamShortWrite if the WriteStream passed to constructor's Write()
+   *       function returns 0.
+   *     - kTvmErrorWriteStreamShortWrite if the WriteStream passed to constructor's Write()
+   *       function returns an invalid positive number.
+   *     - Any negative value (i.e. with bits in kTvmErrorSystemErrorMask set) returned by the
+   *       WriteStream's Write() function.
+   */
```

Comment:
Given rpc_common is quite internal to the uTVM RPC protocol, perhaps we should move them as internal header.